os/arch/arm/src/s5j/s5j_i2s.c

```diff
@@ -1596,117 +1596,128 @@ static uint32_t i2s_txdatawidth(struct i2s_dev_s *dev, int bits)
+		s5j_dmastop(priv->txp.dma);
+
+		while (sq_peek(&priv->txp.pend) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.pend);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_tx_free(priv, bfcontainer);
+		}
 
-	while (sq_peek(&priv->txp.act) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.act);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_tx_free(priv, bfcontainer);
-	}
+		while (sq_peek(&priv->txp.act) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.act);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_tx_free(priv, bfcontainer);
+		}
 
-	while (sq_peek(&priv->txp.done) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.done);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_tx_free(priv, bfcontainer);
+		while (sq_peek(&priv->txp.done) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->txp.done);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_tx_free(priv, bfcontainer);
+		}
 	}
 #endif
 
 #ifdef I2S_HAVE_RX
-	while (sq_peek(&priv->rx.pend) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.pend);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_rx_free(priv, bfcontainer);
-	}
+	if (dir == I2S_RX) {
+		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
+		modifyreg32(priv->base + S5J_I2S_CON, I2S_CR_RXDMACTIVE, 0);
+		s5j_dmastop(priv->rx.dma);
+
+		while (sq_peek(&priv->rx.pend) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.pend);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_rx_free(priv, bfcontainer);
+		}
 
-	s5j_dmastop(priv->rx.dma);
+		while (sq_peek(&priv->rx.act) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.act);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_rx_free(priv, bfcontainer);
+		}
 
-	while (sq_peek(&priv->rx.act) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.act);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_rx_free(priv, bfcontainer);
+		while (sq_peek(&priv->rx.done) != NULL) {
+			flags = irqsave();
+			bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.done);
+			irqrestore(flags);
+			apb_free(bfcontainer->apb);
+			i2s_buf_rx_free(priv, bfcontainer);
+		}
 	}
+#endif
 
-	while (sq_peek(&priv->rx.done) != NULL) {
-		flags = irqsave();
-		bfcontainer = (struct s5j_buffer_s *)sq_remfirst(&priv->rx.done);
-		irqrestore(flags);
-		apb_free(bfcontainer->apb);
-		i2s_buf_rx_free(priv, bfcontainer);
+
+	reg = getreg32(priv->base + S5J_I2S_CON);
+
+	if ((reg & (I2S_CR_TXDMACTIVE | I2S_CR_RXDMACTIVE)) == 0) {
```

Comment:
This condition would disable I2S port when either of RX/TX is NOT active. But I see you want to handle RX & TX seperately. Confirm if this condition is okay.

Reply:
It disables I2S when both RX and TX are not active.